### PR TITLE
Type labels

### PR DESCRIPTION
**Pull request by yafshar**
**Summary**

- Optimize the implementation and store the label map as a C++ unordered_map

A couple of approaches have been tried, and an unordered map is chosen, which has the constant-time complexity for search, insertion, and removal of elements. I ended up keeping the original vector for enough speed up during the write & read of restart files. If memory is an issue, we can remove vectors altogether, but we have to create them for reading and writing each time we need that.

**Author(s)**

Yaser Afshar

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

This PR is backward compatible with the original implementation.

**Implementation Notes**

@jrgissing The examples should be updated.
@jrgissing I think the [write_data](https://github.com/yafshar/lammps/blob/7559cc05361ecfc79038c17e8716e25596a5925a/src/label_map.cpp#L300) should be updated to create an error when Atom Type Labels or any other section is not complete. With the current implementation, and if for any reason Type Labels are incomplete then the output is wrong! it would miss one or more sections.

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Comment by jrgissing** [comment]
thanks for this PR. It looks good, I'll take a closer look soon

> I think the write_data should be updated to create an error when Atom Type Labels or any other section is not complete. With the current implementation, and if for any reason Type Labels are incomplete then the output is wrong! it would miss one or more sections.

Based on discussion with @sjplimp on the main PR, we will allow incomplete label maps. I will add that feature after merging this PR

**Comment by jrgissing** [comment]
so, perhaps I should first ask, is there anything in this update that would prevent incomplete maps?

**Comment by yafshar** [comment]
> so, perhaps I should first ask, is there anything in this update that would prevent incomplete maps?
I do not think so. The map should work for incomplete maps as well.

What do you mean by incomplete maps?  To make sure, do you mean no label to (numeric) type, like below? If it is the case, then there is no problem with this PR. However, if you mean something else, please write an example, and I can try it.
```
C   =>  1
H   =>  2
    =>  3
C2  =>  4
```

**Comment by jrgissing** [comment]
> ```
> C   =>  1
> H   =>  2
>     =>  3
> C2  =>  4
> ```

yes exactly

File: src/atom_vec.cpp [hunk 1]
```diff
@@ -2090,13 +2090,19 @@ int AtomVec::pack_bond(tagint **buf)
...
 void AtomVec::write_bond(FILE *fp, int n, tagint **buf, int index)
 {
-  std::string typestr;
-  for (int i = 0; i < n; i++) {
-    typestr = std::to_string(buf[i][0]);
-    if (atom->types_style == Atom::LABELS)
-      typestr = atom->lmaps[0]->btypelabel[buf[i][0]-1];
-    fmt::print(fp,"{} {} {} {}\n",index,typestr,buf[i][1],buf[i][2]);
-    index++;
+  if (atom->types_style == Atom::LABELS) {
```
[the remark this answers is not in the document]
Comment (yafshar) [reply src/atom_vec.cpp]: I revert this file changes

File: src/label_map.cpp [hunk 2]
```diff
@@ -74,36 +74,44 @@ void LabelMap::modify_lmap(int narg, char **arg)
...
       error->all(FLERR,"Topology type exceeds system topology type");
     std::string slabel(arg[iarg++]);
     if (isdigit(slabel[0]))
       error->all(FLERR,"Type labels cannot start with a number");
-    if (search(slabel,(*labels),ntypes) != -1)
+    if (search(slabel,(*labels_map)) != -1)
       error->all(FLERR,"Type label already exists: types labels must be unique");
-    (*labels)[itype-1] = slabel;
+    std::string &str = (*labels)[itype-1];
+    if (!str.empty()) (*labels_map).erase(str);
```
Comment (jrgissing) [inline src/label_map.cpp:112]: why did you refactor the string assignment here?

Comment (yafshar) [reply src/label_map.cpp:112]: @jrgissing this is done to be compatible with vector implementation. If the `labels` at position `i` is empty means it is the first time assignment and the map also adds the same label. If the user later wants to overwrite the label e.x. with `labelmap atom i re-rename`, if I do not erase the previous label from the map, we would end up with both old and new labels in the map which is not correct.

**Pull request merged**